### ISSUE TITLE
Error discovering fdb-table module

### ISSUE THREAD

**Issue by driesmp** [librenms/librenms#13844]
### The problem

Error discovering fdb-table module. Check log file for more details.

root@librenms:/var/log/librenms # cat librenms.log |grep fdb
[2022-03-11T01:00:03.802435+01:00] production.ERROR: Error discovering fdb-table module for localhost. Illegal offset type
#0 /usr/local/www/librenms/includes/discovery/fdb-table.inc.php(29): include()
[2022-03-11T07:00:02.306179+01:00] production.ERROR: Error discovering fdb-table module for localhost. Illegal offset type
#0 /usr/local/www/librenms/includes/discovery/fdb-table.inc.php(29): include()
[2022-03-11T07:00:03.446301+01:00] production.ERROR: Error discovering fdb-table module for localhost. Illegal offset type
#0 /usr/local/www/librenms/includes/discovery/fdb-table.inc.php(29): include()
[2022-03-11T13:00:03.629126+01:00] production.ERROR: Error discovering fdb-table module for localhost. Illegal offset type
#0 /usr/local/www/librenms/includes/discovery/fdb-table.inc.php(29): include()

### Output of ./validate.php

```txt
root@librenms:/usr/local/www/librenms # su -m www -c "validate.php"
====================================
Component | Version
--------- | -------
LibreNMS  | 22.2.1
DB Schema | 2021_25_01_0129_isis_adjacencies_nullable (234)
PHP       | 8.0.16
Python    |
MySQL     | 10.5.15-MariaDB
RRDTool   | 1.7.2
SNMP      | 5.9
====================================

[OK]    Installed from package; no Composer required
[OK]    Database connection successful
[OK]    Database schema correct
[FAIL]  python3 not found
        [FIX]:
        Install Python 3 for your system.
```

The python error is not true, its because FreeBSD does not link python3 to python3.X. If someone can point me to the code where the python detection is done I'll create a PR.

### What was the last working version of LibreNMS?

21.5.1

### Anything in the logs that might be useful for us?

I'm running FreeBSD 13-STABLE if that's relevant.

**Comment by murrant** [comment]
Seems like something odd with your install.

what is the output of `lnms config:get install_dir`

**Comment by murrant** [comment]
I can't figure out what your error means.

[2022-03-11T13:00:03.629126+01:00] production.ERROR: Error discovering fdb-table module for localhost. Illegal offset type
#0 /usr/local/www/librenms/includes/discovery/fdb-table.inc.php(29): include()

Line 29 of that file is an include, which is consistent with the error, but illegal offset is usually when you try to use null for an array key or something.

I can't decipher or reproduce your issue, so I'm afraid I can't help you.

**Comment by driesmp** [comment]
Is there a way I can run this discovery from the CLI? (maybe its more verbose with errors?)

**Comment by murrant** [comment]
`./discovery.php -h <hostname> -m fdb-table -d`

**Comment by driesmp** [comment]
root@librenms:/usr/local/www/librenms # su -m www -c "php ./discovery.phplocalhost -m fdb-table -d"
LibreNMS Discovery
SQL[select `migration` from `migrations` order by `id` desc limit 1 [] 0.19ms]

SQL[select count(*) as aggregate from `migrations` limit 1 [] 0.12ms]

SQL[select version() [] 0.13ms]


Warning: Undefined array key "local_sha" in /usr/local/www/librenms/discovery.php on line 71
===================================
Version info:
Commit SHA:
Commit Date: 1645175566
DB Schema: 2021_25_01_0129_isis_adjacencies_nullable (234)
PHP: 8.0.17
MySQL: 10.6.7-MariaDB
RRDTool: 1.7.2
SNMP: 5.9
==================================DEBUG!
Updating os_def.cache

Warning: file_put_contents(/usr/local/www/librenms/cache/os_defs.cache): Failed to open stream: No such file or directory in /usr/local/www/librenms/LibreNMS/Util/OS.php on line 111
Override discovery modules: fdb-table
SQL[SELECT * FROM `devices` WHERE disabled = 0 AND `hostname` LIKE 'localhost' ORDER BY device_id DESC [] 0.4ms]

SQL[select * from `devices` where `device_id` = ? limit 1 [6] 0.19ms]

SQL[select * from `devices_attribs` where `devices_attribs`.`device_id` = ? and `devices_attribs`.`device_id` is not null [6] 0.13ms]

localhost 6 freebsd [FPING] '/usr/local/sbin/fping' '-e' '-q' '-c' '3' '-p' '500' '-t' '500' '-O' '0' 'localhost'

response: xmt/rcv/%loss = 3/3/0%, min/avg/max = 0.032/0.066/0.088
SNMP['/usr/local/bin/snmpget' '-M' '/usr/local/www/librenms/mibs' '-v2c' '-c' 'COMMUNITY' '-OQXUte' 'udp:HOSTNAME:161' 'SNMPv2-MIB::sysObjectID.0']
SNMPv2-MIB::sysObjectID.0 = SNMPv2-SMI::enterprises.123*.1.1


Modules status: Global+
OS
Device

#### Load disco module core ####
Attempting to initialize OS: freebsd
Attempting to initialize Group OS: unix
OS initialized: LibreNMS\OS\Shared\Unix
SNMP['/usr/local/bin/snmpget' '-M' '/usr/local/www/librenms/mibs' '-v2c' '-c' 'COMMUNITY' '-OQXUte' '-On' 'udp:HOSTNAME:161' 'SNMPv2-MIB::sysObjectID.0' 'SNMPv2-MIB::sysDescr.0' 'SNMPv2-MIB::sysName.0']
.*.*.0 = .*.4.1.123*.1.1
.*.*.0 = vados.lan 2774037089 FreeBSD 13.1-STABLE
.*.*.0 = vados.lan


| vados.lan 2774037089 FreeBSD 13.1-STABLE | .1.3.6.1.4.1.12325.1.1.2.1.1 |

SNMP['/usr/local/bin/snmpget' '-M' '/usr/local/www/librenms/mibs:/usr/local/www/librenms/mibs/arris' '-v2c' '-c' 'COMMUNITY' '-OQXUte' 'udp:HOSTNAME:161' '.1.3.6.1.4.1.1166.1.621.14.2.0']
SNMPv2-SMI::enterprises.1166.1.6* = No Such Object available on this agent at this OID


SNMP['/usr/local/bin/snmpget' '-M' '/usr/local/www/librenms/mibs' '-v2c' '-c' 'COMMUNITY' '-OQXUte' 'udp:HOSTNAME:161' '.1.3.6.1.4.1.8072.1.3.2.3.1.2.6.100.105.115.116.114.111']
NET-SNMP-EXTEND-MIB::nsExtendOutputFull[distro] = No Such Object available on this agent at this OID


SNMP['/usr/local/bin/snmpget' '-M' '/usr/local/www/librenms/mibs:/usr/local/www/librenms/mibs/dell' '-v2c' '-c' 'COMMUNITY' '-OQXUte' 'udp:HOSTNAME:161' '1.3.6.1.4.1.674.11000.2000.500.1.2.1.0']
SNMPv2-SMI::enterprises.674.11000.2000.5*.0 = No Such Object available on this agent at this OID


SNMP['/usr/local/bin/snmpget' '-M' '/usr/local/www/librenms/mibs:/usr/local/www/librenms/mibs/eltek' '-v2c' '-c' 'COMMUNITY' '-OQXUte' 'udp:HOSTNAME:161' '.1.3.6.1.4.1.12148.10.2.6.0']
SNMPv2-SMI::enterprises.12*.0 = No Such Object available on this agent at this OID


SNMP['/usr/local/bin/snmpget' '-M' '/usr/local/www/librenms/mibs:/usr/local/www/librenms/mibs/hikvision' '-v2c' '-c' 'COMMUNITY' '-OQXUte' 'udp:HOSTNAME:161' '.1.3.6.1.4.1.39165.1.1.0']
SNMPv2-SMI::enterprises.39* = No Such Object available on this agent at this OID


SNMP['/usr/local/bin/snmpget' '-M' '/usr/local/www/librenms/mibs:/usr/local/www/librenms/mibs/orvaldi' '-v2c' '-c' 'COMMUNITY' '-OQXUte' 'udp:HOSTNAME:161' '.1.3.6.1.4.1.21111.1.1.1.6.0']
SNMPv2-SMI::enterprises.21*.6.0 = No Such Object available on this agent at this OID


SNMP['/usr/local/bin/snmpget' '-M' '/usr/local/www/librenms/mibs' '-v2c' '-c' 'COMMUNITY' '-OQXUte' 'udp:HOSTNAME:161' '1.3.6.1.4.1.4515.1.3.6.1.1.1.2.0']
SNMPv2-SMI::enterprises.45*.*.0 = No Such Object available on this agent at this OID


SNMP['/usr/local/bin/snmpget' '-M' '/usr/local/www/librenms/mibs:/usr/local/www/librenms/mibs/wisi' '-v2c' '-c' 'COMMUNITY' '-OQXUte' 'udp:HOSTNAME:161' '.1.3.6.1.4.1.7465.20.2.9.1.2.1.3.1.2']
SNMPv2-SMI::enterprises.74*.*.1.2 = No Such Object available on this agent at this OID


OS: FreeBSD (freebsd)

Attempting to initialize OS: freebsd
Attempting to initialize Group OS: unix
OS initialized: LibreNMS\OS\Shared\Unix

>> Runtime for discovery module 'core': 0.2050 seconds with 276880 bytes
>> SNMP: [9/0.19s] MySQL: [0/0.00s] RRD: [0/0.00s]
#### Unload disco module core ####

Modules status: Global+
OS
Device

#### Load disco module fdb-table ####
SQL[SELECT `vlan_id`, `vlan_vlan` from `vlans` WHERE `device_id` = ? [6] 0.25ms]

SQL[SELECT * FROM `ports_fdb` WHERE `device_id` = ? [6] 0.12ms]

SNMP['/usr/local/bin/snmpbulkwalk' '-v2c' '-c' 'COMMUNITY' '-OQUsetX' '-m' 'Q-BRIDGE-MIB' '-M' '/usr/local/www/librenms/mibs' 'udp:HOSTNAME:161' 'dot1qTpFdbPort']
dot1qTpFdbPort = No Such Object available on this agent at this OID

SNMP['/usr/local/bin/snmpbulkwalk' '-v2c' '-c' 'COMMUNITY' '-OQUsetX' '-m' 'BRIDGE-MIB' '-M' '/usr/local/www/librenms/mibs' 'udp:HOSTNAME:161' 'dot1dTpFdbPort']
dot1dTpFdbPort[6:0:17:10:86:e3].24 = 2
dot1dTpFdbPort[6:36:2c:c4:ca:ec].41 = 2
dot1dTpFdbPort[6:68:63:59:6a:87].214 = 7
dot1dTpFdbPort[6:6a:63:59:6a:87].214 = 7

BRIDGE-MIB: SNMP['/usr/local/bin/snmpbulkwalk' '-v2c' '-c' 'COMMUNITY' '-OQUsetX' '-m' 'BRIDGE-MIB' '-M' '/usr/local/www/librenms/mibs' 'udp:HOSTNAME:161' 'dot1dBasePortIfIndex']
dot1dBasePortIfIndex[2] = 2
dot1dBasePortIfIndex[7] = 7

SQL[SELECT * FROM `ports` WHERE `device_id` = ? AND `ifIndex` = ? [6,"2"] 0.39ms]

SQL[SELECT * FROM `ports` WHERE `device_id` = ? AND `ifIndex` = ? [6,"7"] 0.21ms]

SNMP['/usr/local/bin/snmpbulkwalk' '-v2c' '-c' 'COMMUNITY' '-OQUsetX' '-m' 'Q-BRIDGE-MIB' '-M' '/usr/local/www/librenms/mibs' 'udp:HOSTNAME:161' 'dot1qVlanFdbId']
dot1qVlanFdbId = No Such Object available on this agent at this OID

Error discovering fdb-table module for localhost. Illegal offset type
#0 /usr/local/www/librenms/includes/discovery/fdb-table.inc.php(29): include()
#1 /usr/local/www/librenms/includes/discovery/functions.inc.php(157): include('/usr/local/www/...')
#2 /usr/local/www/librenms/discovery.php(118): discover_device(Array, true)
#3 {main}
SQL[insert into `eventlog` (`device_id`, `reference`, `type`, `datetime`, `severity`, `message`, `username`) values (?, ?, ?, ?, ?, ?, ?) [6,null,"discovery","2022-04-09 10:31:00",5,"Error discovering fdb-table module. Check log file for more details.",""] 1.56ms]


>> Runtime for discovery module 'fdb-table': 0.1240 seconds with 453400 bytes
>> SNMP: [0/0.00s] MySQL: [5/0.03s] RRD: [0/0.00s]
#### Unload disco module fdb-table ####

SQL[UPDATE `devices` set `last_discovered`=NOW(),`last_discovered_timetaken`=? WHERE `device_id` = ? [1.372,6] 1.55ms]

Discovered in 1.372 seconds

./discovery.php localhost 2022-04-09 10:31:00 - 1 devices discovered in 1.526 secs

SNMP [10/0.22s]: Snmpget[10/0.22s]
SQL [15/0.07s]: Select[13/0.04s] Insert[1/0.02s] Update[1/0.02s]
RRD [0/0.00s]:

**Comment by murrant** [comment]
Unfortunately, there is an error that it doesn't show the full error message in the poller output.

Run ./daily.sh (assuming you are using the daily release channel)
and then run it again.

I only need the part that starts with `Error discovering fdb-table module` and all the # lines

**Comment by laf** [comment]
If someone is still having this issue on a more recent code base then please provide a full discovery debug output. `./discovery.php -h HOSTNAME -d -v -m feb-table`

**Comment by cube79** [comment]
@laf I have the problem on one of the devices (the newest one). 

Here's the output from the debug command 
```
administrator@librenms:/opt/librenms$ ./discovery.php -h 192.168.77.81 -d -v -m feb-table
LibreNMS Discovery
SQL[select `migration` from `migrations` order by `id` desc limit 1 [] 0.69ms]

SQL[select count(*) as aggregate from `migrations` [] 0.75ms]

SQL[select version() [] 0.55ms]

===========================================
Component | Version
--------- | -------
LibreNMS  | 25.2.0 (2025-02-21T00:15:22+01:00)
DB Schema | 2025_01_07_223946_drop_cisco_a_s_a_table (315)
PHP       | 8.3.16
Python    | 3.10.12
Database  | MariaDB 10.6.18-MariaDB-0ubuntu0.22.04.1
RRDTool   | 1.7.2
SNMP      | 5.9.1
===========================================
DEBUG!
Reporting disabled by user setting
LibreNMS config cache cleared and config reloaded.
Override discovery modules:
SQL[SELECT * FROM `devices` WHERE disabled = 0 AND `hostname` LIKE '192.168.77.81' ORDER BY device_id DESC [] 0.73ms]

192.168.77.81 84 ciscosb SQL[select * from `devices` where `device_id` = ? limit 1 [84] 0.57ms]

SQL[select * from `devices_attribs` where `devices_attribs`.`device_id` = ? and `devices_attribs`.`device_id` is not null [84] 0.37ms]

[FPING] '/usr/bin/fping' '-e' '-q' '-c' '3' '-p' '500' '-t' '500' '-O' '0' '192.168.77.81'

response: 192.168.77.81 : xmt/rcv/%loss = 3/3/0%, min/avg/max = 0.668/1.15/1.68
SNMP['/usr/bin/snmpget' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/cisco' '-m' 'SNMPv2-TC:SNMPv2-MIB:IF-MIB:IP-MIB:TCP-MIB:UDP-MIB:NET-SNMP-VACM-MIB' '-v2c' '-c' 'Comm-string' '-OQXUte' '-Pu' 'udp:192.168.77.81:161' 'SNMPv2-MIB::sysObjectID.0']
SNMPv2-MIB::sysObjectID.0 = SNMPv2-SMI::enterprises.9.6.1.1010.24.5


Modules status: Global+
OS
Device

#### Load disco module core ####
Attempting to initialize OS: ciscosb
OS initialized: LibreNMS\OS\Ciscosb
SNMP['/usr/bin/snmpget' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/cisco' '-m' 'SNMPv2-TC:SNMPv2-MIB:IF-MIB:IP-MIB:TCP-MIB:UDP-MIB:NET-SNMP-VACM-MIB' '-v2c' '-c' 'Comm-string' '-OQXUte' '-Pu' '-On' 'udp:192.168.77.81:161' 'SNMPv2-MIB::sysObjectID.0' 'SNMPv2-MIB::sysDescr.0' 'SNMPv2-MIB::sysName.0']
.1.3.6.1.2.1.1.2.0 = .1.3.6.1.4.1.9.6.1.1010.24.5
.1.3.6.1.2.1.1.1.0 = 24-Port Gigabit PoE Smart Switch
.1.3.6.1.2.1.1.5.0 = Switch4ED523


| 24-Port Gigabit PoE Smart Switch | .1.3.6.1.4.1.9.6.1.1010.24.5 |

OS: Cisco Small Business (ciscosb)

Attempting to initialize OS: ciscosb
OS initialized: LibreNMS\OS\Ciscosb

>> Runtime for discovery module 'core': 0.0830 seconds with 374080 bytes
>> SNMP: [1/0.06s] MySQL: [0/0.00s] RRD: [0/0.00s]
#### Unload disco module core ####

SQL[update `devices` set `last_discovered_timetaken` = ?, `last_discovered` = NOW() where `device_id` = ? [1.133,84] 828.99ms]

Discovered in 1.133 seconds

./discovery.php 192.168.77.81 2025-02-28 07:10:46 - 1 devices discovered in 2.293 secs

SNMP [2/0.09s]: Snmpget[2/0.09s]
SQL [8/8.35s]: Select[7/0.06s] Update[1/8.29s]
RRD [0/0.00s]:
```

**Comment by murrant** [comment]
Please start a new bug if issues persist
